cabal-install/src/Distribution/Client/CmdInstall.hs

```diff
@@ -355,7 +357,10 @@ installAction flags@NixStyleFlags { extraFlags = clientInstallFlags', .. } targe
   envFile <- getEnvFile clientInstallFlags platform compilerVersion
   existingEnvEntries <-
     getExistingEnvEntries verbosity compilerFlavor supportsPkgEnvFiles envFile
-  packageDbs <- getPackageDbStack compilerId projectConfigStoreDir projectConfigLogsDir
+  packageDbs' <- getPackageDbStack compilerId projectConfigStoreDir projectConfigLogsDir
+  let validDb (SpecificPackageDB fp) = doesFileExist fp
+      validDb _ = pure True
+  packageDbs <- filterM validDb packageDbs'
```

Comment:
Also, the implementation of `validDb` is incorrect because a package DB is a directory (I'm not sure it is specified to be a directory, but it was one when I ran a debug).  It should be:

```Haskell
validDb (SpecificPackageDB fp) = doesPathExist {- maybe `doesDirectoryExist` ? -} fp
validDb _ = pure True
```

Reply:
It looks like the two points above are actually a part of the cause of the problem. See the following shell session:

```console
$ cd ~/Downloads/cabal/

$ git status
On branch gb/base-in-cabal-install-lib
Your branch is up to date with 'origin/gb/base-in-cabal-install-lib'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   cabal-install/src/Distribution/Client/CmdInstall.hs

no changes added to commit (use "git add" and/or "git commit -a")

$ git diff
diff --git a/cabal-install/src/Distribution/Client/CmdInstall.hs b/cabal-install/src/Distribution/Client/CmdInstall.hs
index 49b264f36..8ed288028 100644
--- a/cabal-install/src/Distribution/Client/CmdInstall.hs
+++ b/cabal-install/src/Distribution/Client/CmdInstall.hs
@@ -358,8 +358,13 @@ installAction flags@NixStyleFlags { extraFlags = clientInstallFlags', .. } targe
   existingEnvEntries <-
     getExistingEnvEntries verbosity compilerFlavor supportsPkgEnvFiles envFile
   packageDbs' <- getPackageDbStack compilerId projectConfigStoreDir projectConfigLogsDir
-  let validDb (SpecificPackageDB fp) = doesFileExist fp
-      validDb _ = pure True
+  let validDb' (SpecificPackageDB fp) = doesPathExist fp
+      validDb' _ = pure True
+      validDb db = do
+        putStrLn $ "validDb called against " ++ show db
+        res <- validDb' db
+        putStrLn $ "validDb res: " ++ show res
+        return res
   packageDbs <- filterM validDb packageDbs'
   installedIndex <- getInstalledPackages verbosity compiler packageDbs progDb
 
$ cabal v2-install --installdir=$HOME --install-method=copy --project-file=cabal.project.release cabal-install --overwrite-policy=always
Wrote tarball sdist to
/Users/gksato/Downloads/cabal/dist-newstyle/sdist/Cabal-syntax-3.11.0.0.tar.gz
Wrote tarball sdist to
/Users/gksato/Downloads/cabal/dist-newstyle/sdist/Cabal-3.11.0.0.tar.gz
Wrote tarball sdist to
/Users/gksato/Downloads/cabal/dist-newstyle/sdist/cabal-install-solver-3.11.0.0.tar.gz
Wrote tarball sdist to
/Users/gksato/Downloads/cabal/dist-newstyle/sdist/cabal-install-3.11.0.0.tar.gz
Resolving dependencies...
Up to date
Copying 'cabal' to '/Users/gksato/cabal'

$ cd ~
$ mv ~/.cabal ~/.cabal-save
$ rm -rf ~/.cabal ~/.cache/cabal ~/.cache/cabal-helper ~/.config/cabal ~/.ghc ~/.local/state/cabal

$ ~/cabal v2-update
Config file path source is default config file.
Config file not found: /Users/gksato/.config/cabal/config
Writing default configuration to /Users/gksato/.config/cabal/config
Downloading the latest package list from hackage.haskell.org
Package list of hackage.haskell.org has been updated.
The index-state is set to 2023-04-23T19:32:13Z.

$ ~/cabal v2-install --lib leancheck
validDb called against GlobalPackageDB
validDb res: True
validDb called against SpecificPackageDB "/Users/gksato/.local/state/cabal/store/ghc-9.4.4/package.db"
validDb res: False
Resolving dependencies...
Build profile: -w ghc-9.4.4 -O1
In order, the following will be built (use -v for more details):
 - leancheck-1.0.0 (lib) (requires download & build)
Downloading  leancheck-1.0.0
Downloaded   leancheck-1.0.0
Starting     leancheck-1.0.0 (lib)
Building     leancheck-1.0.0 (lib)
Installing   leancheck-1.0.0 (lib)
Completed    leancheck-1.0.0 (lib)

$ cat ~/.ghc/x86_64-darwin-9.4.4/environments/default 
clear-package-db
global-package-db
package-id base-4.17.0.0
package-id lnchck-1.0.0-b81085f1

$ ghci
Loaded package environment from /Users/gksato/.ghc/x86_64-darwin-9.4.4/environments/default
GHCi, version 9.4.4: https://www.haskell.org/ghc/  :? for help
<command line>: cannot satisfy -package-id lnchck-1.0.0-b81085f1
    (use -v for more information)

$ rm -rf ~/.ghc

$ ~/cabal v2-install --lib leancheck
validDb called against GlobalPackageDB
validDb res: True
validDb called against SpecificPackageDB "/Users/gksato/.local/state/cabal/store/ghc-9.4.4/package.db"
validDb res: True
Resolving dependencies...

$ cat ~/.ghc/x86_64-darwin-9.4.4/environments/default 
clear-package-db
global-package-db
package-db /Users/gksato/.local/state/cabal/store/ghc-9.4.4/package.db
package-id base-4.17.0.0
package-id lnchck-1.0.0-b81085f1

$ ghci
Loaded package environment from /Users/gksato/.ghc/x86_64-darwin-9.4.4/environments/default
GHCi, version 9.4.4: https://www.haskell.org/ghc/  :? for help
ghci> putStrLn "GHCi ran successfully!"
GHCi ran successfully!
ghci> :q
Leaving GHCi.
```

Reply:
thanks for the careful investigation! Both issues seem legit and I'll try to do some more work in the next few days...

Reply:
Fixed both issues raised and tested to ensure they covered the problem. Thanks again!